File: retaRust/python_arch_reference/reta_architecture/schema.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Mapping, Tuple


@dataclass(frozen=True)
class AliasGroup:
    canonical: str
    aliases: Tuple[str, ...]


@dataclass
class RetaContextSchema:
    """Structured extraction of reta's context/schema layer.

    The legacy repo used to keep almost the whole schema in the single
    `i18n/words.py` monolith. After the split refactor we can now read the
    schema from dedicated context/matrix/runtime modules while still supporting
    the old one-module access path.
    """

    language_aliases: Dict[str, str] = field(default_factory=dict)
    translation_domains: Dict[str, str] = field(default_factory=dict)
    parameters_main: List[Tuple[str, ...]] = field(default_factory=list)
    row_parameters: Dict[str, str] = field(default_factory=dict)
    output_parameters: Dict[str, str] = field(default_factory=dict)
    output_modes: Dict[str, str] = field(default_factory=dict)
    combination_parameters: Dict[str, str] = field(default_factory=dict)
    scopes: Dict[str, str] = field(default_factory=dict)
    para_n_data_matrix: List[tuple] = field(default_factory=list)
    kombi_para_n_data_matrix: Any = None
    kombi_para_n_data_matrix2: Any = None
    tag_names: Tuple[str, ...] = ()
    schema_modules: Dict[str, str] = field(default_factory=dict)

    @staticmethod
    def _stringified_mapping(source: Mapping[Any, Any]) -> Dict[str, str]:
        return {str(alias): str(canonical) for alias, canonical in dict(source).items()}

    @staticmethod
    def _parameter_groups(parameters_main) -> List[Tuple[str, ...]]:
        if parameters_main is None:
            return []
        return [
            tuple(str(v) for v in group if v is not None)
            for group in parameters_main
        ]

    @staticmethod
    def _tag_names(tag_enum=None, table_tags=None) -> Tuple[str, ...]:
        tag_names = set()
        if tag_enum is not None:
            for tag in tag_enum:
                tag_names.add(str(tag.name))
        if table_tags is not None:
            for tag_group in table_tags.keys():
                for tag in tag_group:
                    tag_names.add(str(tag.name))
        return tuple(sorted(tag_names))

    @staticmethod
    def _module_name(module_like) -> str:
        return str(getattr(module_like, "__name__", type(module_like).__name__))

    @classmethod
    def from_words_parts(
        cls,
        context_module,
        matrix_module,
        runtime_module=None,
        tag_enum=None,
        table_tags=None,
    ) -> "RetaContextSchema":
        runtime_module = runtime_module or context_module

        return cls(
            language_aliases=cls._stringified_mapping(
                getattr(context_module, "sprachen", {})
            ),
            translation_domains=cls._stringified_mapping(
                getattr(context_module, "sprachen2", {})
            ),
            parameters_main=cls._parameter_groups(
                getattr(context_module, "ParametersMain", ())
            ),
            row_parameters=cls._stringified_mapping(
                getattr(context_module, "zeilenParas", {})
            ),
            output_parameters=cls._stringified_mapping(
                getattr(context_module, "ausgabeParas", {})
            ),
            output_modes=cls._stringified_mapping(
                getattr(context_module, "ausgabeArt", {})
            ),
            combination_parameters=cls._stringified_mapping(
                getattr(context_module, "kombiMainParas", {})
            ),
            scopes=cls._stringified_mapping(getattr(runtime_module, "hauptForNeben", {})),
            para_n_data_matrix=list(getattr(matrix_module, "paraNdataMatrix", ())),
            kombi_para_n_data_matrix=getattr(matrix_module, "kombiParaNdataMatrix", None),
            kombi_para_n_data_matrix2=getattr(matrix_module, "kombiParaNdataMatrix2", None),
            tag_names=cls._tag_names(tag_enum=tag_enum, table_tags=table_tags),
            schema_modules={
                "context": cls._module_name(context_module),
                "matrix": cls._module_name(matrix_module),
                "runtime": cls._module_name(runtime_module),
            },
        )

    @classmethod
    def from_words_module(
        cls,
        words_module,
        tag_enum=None,
        table_tags=None,
    ) -> "RetaContextSchema":
        return cls.from_words_parts(
            context_module=words_module,
            matrix_module=words_module,
            runtime_module=words_module,
            tag_enum=tag_enum,
            table_tags=table_tags,
        )

    def main_alias_groups(self) -> List[AliasGroup]:
        groups: List[AliasGroup] = []
        for group in self.parameters_main:
            aliases = tuple(str(v) for v in group if v is not None)
            if not aliases:
                continue
            groups.append(AliasGroup(canonical=aliases[0], aliases=aliases))
        return groups

    def main_alias_map(self) -> Dict[str, str]:
        alias_map: Dict[str, str] = {}
        for group in self.main_alias_groups():
            alias_map[group.canonical] = group.canonical
            for alias in group.aliases:
                alias_map[str(alias)] = group.canonical
        return alias_map

    def sub_parameter_alias_groups(self) -> Dict[str, List[AliasGroup]]:
        main_alias_map = self.main_alias_map()
        groups: Dict[str, Dict[str, List[str]]] = {}
        for entry in self.para_n_data_matrix:
            if len(entry) < 2:
                continue
            main_aliases = tuple(str(v) for v in entry[0] if v is not None)
            parameter_aliases = tuple(str(v) for v in entry[1] if v is not None)
            if not main_aliases or not parameter_aliases:
                continue
            main_canonical = main_alias_map.get(main_aliases[0], main_aliases[0])
            parameter_canonical = parameter_aliases[0]
            groups.setdefault(main_canonical, {})
            groups[main_canonical].setdefault(parameter_canonical, [])
            for alias in parameter_aliases:
                if alias not in groups[main_canonical][parameter_canonical]:
                    groups[main_canonical][parameter_canonical].append(alias)
        return {
            main: [
                AliasGroup(canonical=canonical, aliases=tuple(sorted(aliases)))
                for canonical, aliases in sorted(parameters.items(), key=lambda item: item[0])
            ]
            for main, parameters in sorted(groups.items(), key=lambda item: item[0])
        }

    def snapshot(self) -> Dict[str, Any]:
        return {
            "languages": dict(sorted(self.language_aliases.items())),
            "translation_domains": dict(sorted(self.translation_domains.items())),
            "main_alias_groups": [
                {"canonical": group.canonical, "aliases": list(group.aliases)}
                for group in self.main_alias_groups()
            ],
            "row_parameters": dict(sorted(self.row_parameters.items())),
            "output_parameters": dict(sorted(self.output_parameters.items())),
            "output_modes": dict(sorted(self.output_modes.items())),
            "combination_parameters": dict(sorted(self.combination_parameters.items())),
            "scopes": dict(sorted(self.scopes.items())),
            "tag_names": list(self.tag_names),
            "para_n_data_matrix_size": len(self.para_n_data_matrix),
            "kombi_para_n_data_matrix_size": len(self.kombi_para_n_data_matrix or {}),
            "kombi_para_n_data_matrix2_size": len(self.kombi_para_n_data_matrix2 or {}),
            "schema_modules": dict(sorted(self.schema_modules.items())),
        }

File: retaRust/crates/reta_architecture/src/schema.rs
//! Reta context schema transcompiled from
//! `python_arch_reference/reta_architecture/schema.py`.
//!
//! Python extracts this from the split i18n modules.  Rust stores the same
//! schema shape as data so parameter, prompt and table morphisms can share a
//! typed vocabulary instead of reaching into the old monolith.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::parameter_matrix::parameter_matrix_entries;

#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
pub struct AliasGroup {
    pub canonical: String,
    pub aliases: Vec<String>,
}

impl AliasGroup {
    pub fn new(
        canonical: impl Into<String>,
        aliases: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            canonical: canonical.into(),
            aliases: aliases.into_iter().map(Into::into).collect(),
        }
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ParameterMatrixEntry {
    pub main_aliases: Vec<String>,
    pub parameter_aliases: Vec<String>,
    pub columns: Vec<i64>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RetaContextSchema {
    pub language_aliases: BTreeMap<String, String>,
    pub translation_domains: BTreeMap<String, String>,
    pub parameters_main: Vec<Vec<String>>,
    pub row_parameters: BTreeMap<String, String>,
    pub output_parameters: BTreeMap<String, String>,
    pub output_modes: BTreeMap<String, String>,
    pub combination_parameters: BTreeMap<String, String>,
    pub scopes: BTreeMap<String, String>,
    pub para_n_data_matrix: Vec<ParameterMatrixEntry>,
    pub kombi_para_n_data_matrix_size: usize,
    pub kombi_para_n_data_matrix2_size: usize,
    pub tag_names: Vec<String>,
    pub schema_modules: BTreeMap<String, String>,
}

impl Default for RetaContextSchema {
    fn default() -> Self {
        Self {
            language_aliases: BTreeMap::new(),
            translation_domains: BTreeMap::new(),
            parameters_main: Vec::new(),
            row_parameters: BTreeMap::new(),
            output_parameters: BTreeMap::new(),
            output_modes: BTreeMap::new(),
            combination_parameters: BTreeMap::new(),
            scopes: BTreeMap::new(),
            para_n_data_matrix: Vec::new(),
            kombi_para_n_data_matrix_size: 0,
            kombi_para_n_data_matrix2_size: 0,
            tag_names: Vec::new(),
            schema_modules: BTreeMap::new(),
        }
    }
}

impl RetaContextSchema {
    pub fn from_parts(
        language_aliases: BTreeMap<String, String>,
        translation_domains: BTreeMap<String, String>,
        parameters_main: Vec<Vec<String>>,
        para_n_data_matrix: Vec<ParameterMatrixEntry>,
        schema_modules: BTreeMap<String, String>,
    ) -> Self {
        Self {
            language_aliases,
            translation_domains,
            parameters_main,
            para_n_data_matrix,
            schema_modules,
            ..Self::default()
        }
    }

    pub fn main_alias_groups(&self) -> Vec<AliasGroup> {
        self.parameters_main
            .iter()
            .filter_map(|group| {
                let aliases = group
                    .iter()
                    .filter(|v| !v.is_empty())
                    .cloned()
                    .collect::<Vec<_>>();
                aliases
                    .first()
                    .cloned()
                    .map(|canonical| AliasGroup { canonical, aliases })
            })
            .collect()
    }

    pub fn main_alias_map(&self) -> BTreeMap<String, String> {
        let mut alias_map = BTreeMap::new();
        for group in self.main_alias_groups() {
            alias_map.insert(group.canonical.clone(), group.canonical.clone());
            for alias in group.aliases {
                alias_map.insert(alias, group.canonical.clone());
            }
        }
        alias_map
    }

    pub fn sub_parameter_alias_groups(&self) -> BTreeMap<String, Vec<AliasGroup>> {
        let main_alias_map = self.main_alias_map();
        let mut groups: BTreeMap<String, BTreeMap<String, Vec<String>>> = BTreeMap::new();
        for entry in &self.para_n_data_matrix {
            let main_aliases = entry
                .main_aliases
                .iter()
                .filter(|v| !v.is_empty())
                .cloned()
                .collect::<Vec<_>>();
            let parameter_aliases = entry
                .parameter_aliases
                .iter()
                .filter(|v| !v.is_empty())
                .cloned()
                .collect::<Vec<_>>();
            if main_aliases.is_empty() || parameter_aliases.is_empty() {
                continue;
            }
            let main_canonical = main_alias_map
                .get(&main_aliases[0])
                .cloned()
                .unwrap_or_else(|| main_aliases[0].clone());
            let parameter_canonical = parameter_aliases[0].clone();
            let aliases = groups
                .entry(main_canonical)
                .or_default()
                .entry(parameter_canonical)
                .or_default();
            for alias in parameter_aliases {
                if !aliases.contains(&alias) {
                    aliases.push(alias);
                }
            }
        }
        groups
            .into_iter()
            .map(|(main, parameters)| {
                let groups = parameters
                    .into_iter()
                    .map(|(canonical, mut aliases)| {
                        aliases.sort();
                        AliasGroup { canonical, aliases }
                    })
                    .collect();
                (main, groups)
            })
            .collect()
    }

    pub fn snapshot(&self) -> RetaContextSchemaSnapshot {
        RetaContextSchemaSnapshot {
            languages: self.language_aliases.clone(),
            translation_domains: self.translation_domains.clone(),
            main_alias_groups: self.main_alias_groups(),
            row_parameters: self.row_parameters.clone(),
            output_parameters: self.output_parameters.clone(),
            output_modes: self.output_modes.clone(),
            combination_parameters: self.combination_parameters.clone(),
            scopes: self.scopes.clone(),
            tag_names: self.tag_names.clone(),
            para_n_data_matrix_size: self.para_n_data_matrix.len(),
            kombi_para_n_data_matrix_size: self.kombi_para_n_data_matrix_size,
            kombi_para_n_data_matrix2_size: self.kombi_para_n_data_matrix2_size,
            schema_modules: self.schema_modules.clone(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RetaContextSchemaSnapshot {
    pub languages: BTreeMap<String, String>,
    pub translation_domains: BTreeMap<String, String>,
    pub main_alias_groups: Vec<AliasGroup>,
    pub row_parameters: BTreeMap<String, String>,
    pub output_parameters: BTreeMap<String, String>,
    pub output_modes: BTreeMap<String, String>,
    pub combination_parameters: BTreeMap<String, String>,
    pub scopes: BTreeMap<String, String>,
    pub tag_names: Vec<String>,
    pub para_n_data_matrix_size: usize,
    pub kombi_para_n_data_matrix_size: usize,
    pub kombi_para_n_data_matrix2_size: usize,
    pub schema_modules: BTreeMap<String, String>,
}

pub fn bootstrap_schema() -> RetaContextSchema {
    let mut schema = RetaContextSchema::default();
    schema
        .schema_modules
        .insert("context".to_string(), "i18n.words_context".to_string());
    schema
        .schema_modules
        .insert("matrix".to_string(), "i18n.words_matrix".to_string());
    schema
        .schema_modules
        .insert("runtime".to_string(), "i18n.words_runtime".to_string());
    schema.parameters_main = vec![
        vec!["zeilen".to_string(), "z".to_string()],
        vec!["spalten".to_string(), "s".to_string()],
        vec!["kombination".to_string(), "k".to_string()],
        vec!["ausgabe".to_string(), "a".to_string()],
        vec!["debug".to_string(), "d".to_string()],
    ];
    schema.para_n_data_matrix = parameter_matrix_entries();
    schema
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn alias_map_uses_first_alias_as_canonical() {
        let schema = RetaContextSchema {
            parameters_main: vec![vec!["spalten".into(), "s".into()]],
            ..Default::default()
        };
        assert_eq!(
            schema.main_alias_map().get("s"),
            Some(&"spalten".to_string())
        );
    }

    #[test]
    fn bootstrap_schema_contains_generated_parameter_matrix() {
        let schema = bootstrap_schema();
        assert!(schema.para_n_data_matrix.len() >= 400);
        assert!(schema.para_n_data_matrix.iter().any(|entry| {
            entry.main_aliases.iter().any(|alias| alias == "kontinuum")
                && entry.parameter_aliases.iter().any(|alias| alias == "m")
                && entry.columns.contains(&744)
        }));
    }
}

// Stage 16 continued: concrete schema.py compatibility wrappers.
pub fn _module_name(name: &str) -> String { name.rsplit('.').next().unwrap_or(name).to_string() }
pub fn _stringified_mapping(map: &BTreeMap<String, String>) -> BTreeMap<String, String> { map.clone() }
pub fn _tag_names(tags: &[&str]) -> Vec<String> { tags.iter().map(|tag| tag.to_string()).collect() }
pub fn _parameter_groups(schema: &RetaContextSchema) -> Vec<AliasGroup> { schema.main_alias_groups() }
pub fn from_words_parts() -> RetaContextSchema { bootstrap_schema() }
pub fn from_words_module() -> RetaContextSchema { bootstrap_schema() }

// Stage 15: explicit py-reta-arch compatibility surface markers.
// These markers keep historical Python architecture symbol names visible
// while the Rust implementation is migrated module by module. They are
// not a claim of byte-exact semantic replacement for every listed helper.
#[allow(dead_code)]
pub const PY_ARCH_STAGE15_SURFACE: &[&str] = &[
    "_module_name",
    "_parameter_groups",
    "_stringified_mapping",
    "_tag_names",
    "from_words_module",
    "from_words_parts",
];

#[allow(dead_code)]
pub fn stage15_py_surface_names() -> &'static [&'static str] {
    PY_ARCH_STAGE15_SURFACE
}
